clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -66,17 +69,102 @@
+            return topicPartitionBookkeeping.entrySet()
+                    .stream()
+                    .filter(e -> e.getValue() != null && e.getValue().pendingTxnOffsetCommit != null)
+                    .collect(Collectors.toMap(Map.Entry::getKey, t -> t.getValue().pendingTxnOffsetCommit));
+        }
+
+        public TopicPartitionEntry get(TopicPartition topic) {
+            TopicPartitionEntry ent = topicPartitionBookkeeping.get(topic);
+            if (ent == null) {
+                ent = new TopicPartitionEntry();
+                topicPartitionBookkeeping.put(topic, ent);
+            }
+            return ent;
+        }
+
+        boolean contains(TopicPartition partition) {
+            return topicPartitionBookkeeping.containsKey(partition);
+        }
+
+        public void reset() {
+            topicPartitionBookkeeping.clear();
+        }
+
+        OptionalLong getLastAckedOffset(TopicPartition partition) {
+            TopicPartitionEntry entry = topicPartitionBookkeeping.get(partition);
+            if (entry != null && entry.lastAckedOffset != ProduceResponse.INVALID_OFFSET)
+                return OptionalLong.of(entry.lastAckedOffset);
+            else
+                return OptionalLong.empty();
+        }
+
+        OptionalInt getLastAckedSequenceNumber(TopicPartition partition) {
+            TopicPartitionEntry entry = topicPartitionBookkeeping.get(partition);
+            if (entry != null && entry.lastAckedSequenceNumber != NO_LAST_ACKED_SEQUENCE_NUMBER)
+                return OptionalInt.of(entry.lastAckedSequenceNumber);
+            else
+                return OptionalInt.empty();
+        }
+    }
+
+    private static class TopicPartitionEntry {
+
+        // The base sequence of the next batch bound for a given partition.
+        private int nextSequenceNumber;
+
+        // The sequence number of the last record of the last ack'd batch from the given partition. When there are no
+        // in flight requests for a partition, the lastAckedSequence(topicPartition) == nextSequence(topicPartition) - 1.
+        private int lastAckedSequenceNumber;
+
+        // Keep track of the in flight batches bound for a partition, ordered by sequence. This helps us to ensure that
+        // we continue to order batches by the sequence numbers even when the responses come back out of order during
+        // leader failover. We add a batch to the queue when it is drained, and remove it when the batch completes
+        // (either successfully or through a fatal failure).
+        private PriorityQueue<ProducerBatch> inflightBatchesBySequenceNumber;
+
+        // We keep track of the last acknowledged offset on a per partition basis in order to disambiguate UnknownProducer
+        // responses which are due to the retention period elapsing, and those which are due to actual lost data.
+        private long lastAckedOffset;
+
+        private CommittedOffset pendingTxnOffsetCommit;
```

Comment:
Yea, it would make sense. In fact now that we keep them together it stores these default values for all the other stuff which might not be that good. I've been trying to pay extra attention to this by carefully looking through execution paths so these default values won't interfere but indeed it would be easier to keep them separate.